editoast/src/models/rolling_stock_model.rs

```diff
@@ -138,53 +138,78 @@ impl From<model::Error> for Error {
-                self.raise_pantograph_time.flatten().map(units::second::new),
-            )
-            .map_err(|e| {
-                let mut err = ValidationErrors::new();
-                err.add("effort_curves", e);
-                err
-            }),
-            None => {
-                let mut err = ValidationErrors::new();
-                err.add(
-                    "effort_curves",
-                    ValidationError::new("effort_curves is required"),
-                );
-                Err(err)
+                    .map(units::second::new)
+                    .is_none()
+                {
+                    let mut error = ValidationError::new("electrical_power_startup_time");
+                    error.message = Some(
+                        "electrical_power_startup_time is required for electric rolling stocks"
+                            .into(),
+                    );
+                    validation_errors.add("effort_curves", error);
+                }
+                if self
+                    .raise_pantograph_time
+                    .flatten()
+                    .map(units::second::new)
+                    .is_none()
+                {
+                    let mut error = ValidationError::new("raise_pantograph_time");
+                    error.message = Some(
+                        "raise_pantograph_time is required for electric rolling stocks".into(),
+                    );
+                    validation_errors.add("effort_curves", error);
+                }
             }
+        } else {
+            validation_errors.add(
+                "effort_curves",
+                ValidationError::new("effort_curves is required"),
+            );
         }
     }
-}
 
-pub fn validate_rolling_stock(
-    effort_curves: &EffortCurves,
-    electrical_power_startup_time: Option<Time>,
-    raise_pantograph_time: Option<Time>,
-) -> std::result::Result<(), ValidationError> {
-    if !effort_curves.is_electric() {
-        return Ok(());
-    }
-    if electrical_power_startup_time.is_none() {
-        let mut error = ValidationError::new("electrical_power_startup_time");
-        error.message =
-            Some("electrical_power_startup_time is required for electric rolling stocks".into());
-        return Err(error);
-    }
-    if raise_pantograph_time.is_none() {
-        let mut error = ValidationError::new("raise_pantograph_time");
-        error.message =
-            Some("raise_pantograph_time is required for electric rolling stocks".into());
-        return Err(error);
+    fn validate_primary_category(&self, validation_errors: &mut ValidationErrors) {
+        if let Some(primary_category) = &self.primary_category {
+            if let Some(other_categories) = &self.other_categories {
+                if other_categories
+                    .iter()
+                    .flatten()
+                    .collect::<Vec<_>>()
+                    .contains(&primary_category)
```

Comment:
You should be able to use [`.any()`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.any) to avoid collecting?

```suggestion
                if other_categories
                    .iter()
                    .flatten()
                    .any(|category| category == &primary_category)
```

Reply:
Done.